spv/sync.go

```diff
@@ -0,0 +1,1420 @@
+		}()
+
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		time.Sleep(5 * time.Second)
+	}
+}
+
+func (s *Syncer) connectToCandidates(ctx context.Context) error {
+	var wg sync.WaitGroup
+	defer wg.Wait()
+
+	sem := make(chan struct{}, 8)
+	for {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		select {
+		case sem <- struct{}{}:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+		na, err := s.peerCandidate(reqSvcs)
+		if err != nil {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-time.After(5 * time.Second):
+				<-sem
+				continue
+			}
+		}
+
+		wg.Add(1)
+		go func() {
+			ctx, cancel := context.WithCancel(ctx)
+			defer func() {
+				cancel()
+				wg.Done()
+				<-sem
+			}()
+
+			// Make outbound connections to remote peers.
+			port := strconv.FormatUint(uint64(na.Port), 10)
+			raddr := net.JoinHostPort(na.IP.String(), port)
+			k := addrmgr.NetAddressKey(na)
+
+			s.remotesMu.Lock()
+			s.connectingRemotes[k] = struct{}{}
+			s.remotesMu.Unlock()
+
+			rp, err := s.lp.ConnectOutbound(ctx, raddr, reqSvcs)
+			if err != nil {
+				s.remotesMu.Lock()
+				delete(s.connectingRemotes, k)
+				s.remotesMu.Unlock()
+				if ctx.Err() == nil {
+					log.Warnf("Peering attempt failed: %v", err)
+				}
+				return
+			}
+			log.Infof("New peer %v %v %v", raddr, rp.UA(), rp.Services())
+
+			s.remotesMu.Lock()
+			delete(s.connectingRemotes, k)
+			s.remotes[k] = rp
+			n := len(s.remotes)
+			s.remotesMu.Unlock()
+			s.peerConnected(n, k)
+
+			wait := make(chan struct{})
+			go func() {
+				err := s.startupSync(ctx, rp)
+				if err != nil {
+					rp.Disconnect(err)
+				}
+				wait <- struct{}{}
+			}()
+
+			err = rp.Err()
+			if ctx.Err() != context.Canceled {
+				log.Warnf("Lost peer %v: %v", raddr, err)
+			}
+
+			<-wait
+			s.remotesMu.Lock()
+			delete(s.remotes, k)
+			n = len(s.remotes)
+			s.remotesMu.Unlock()
+			s.peerDisconnected(n, k)
+		}()
+	}
+}
+
+func (s *Syncer) forRemotes(f func(rp *p2p.RemotePeer) error) error {
+	defer s.remotesMu.Unlock()
+	s.remotesMu.Lock()
+	if len(s.remotes) == 0 {
+		return errors.E(errors.NoPeers)
+	}
+	for _, rp := range s.remotes {
+		err := f(rp)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Syncer) pickRemote(pick func(*p2p.RemotePeer) bool) (*p2p.RemotePeer, error) {
+	defer s.remotesMu.Unlock()
+	s.remotesMu.Lock()
+
+	for _, rp := range s.remotes {
+		if pick(rp) {
+			return rp, nil
+		}
+	}
+	return nil, errors.E(errors.NoPeers)
+}
+
+func (s *Syncer) getTransactionsByHashes(ctx context.Context, txHashes []*chainhash.Hash) ([]*wire.MsgTx, []*wire.InvVect, error) {
+	if len(txHashes) == 0 {
+		return nil, nil, nil
+	}
+
+	var notFound []*wire.InvVect
+	var foundTxs []*wire.MsgTx
+
+	for walletID, w := range s.wallets {
+		var missingWalletTxs []*wire.InvVect
+		var foundWalletTxs []*wire.MsgTx
+		var err error
+		foundWalletTxs, missingWalletTxs, err = w.GetTransactionsByHashes(ctx, txHashes)
```

Comment:
The declarations in the 3 lines above this line are unnecessary. Can just be:
```suggestion
		walletFoundTxs, walletNotFoundTxs, err := w.GetTransactionsByHashes(ctx, txHashes)
```
Notice also the recommended variable name changes.